### ISSUE TITLE
Adding YellowFin to the lasagne/optimizers

### ISSUE THREAD

**Issue by partobs-mdp** [Lasagne/Lasagne#856]
YellowFin is a novel method for tuning momentum hyperparameters (learning rate & momentum). It blends the nice sides of momentum SGD (convergence properties) and Adam and similar (automatic tuning of the hyperparameters, in contrast to hand-tuning them).

The original paper (https://arxiv.org/abs/1706.03471) shows some nice convergence speed-ups (2.5x on CIFAR against Adam, for example). I would like to start implementing it as soon as #843 will be resolved.

**Comment by botev** [comment]
Just for anyone interested I made a gist for it, which pretty much based on Lasagne:
https://gist.github.com/botev/f8b32c00eafee222e47393f7f0747666

**Comment by partobs-mdp** [comment]
```python
    # Both times inverse of sigma p
    kl /= T.sqr(1)
    # Minus 1
    kl -= T.constant(1)
    # Log determinant of p_sigma
    kl += 2 * T.log(1)
```

Are you sure you wanted to use three constant terms? Maybe you meant `T.log(p_sigma)` in the last case?

**Comment by botev** [comment]
Ah that is true, but in the case, it's a VAE with 1 layer so `mu_sigma=0` and `p_sigma=1`. It is a copy-pasted code from my own repo and I did not delete the comments.

**Comment by f0k** [comment]
Thank you for the proposal, and for the gist! I agree with @botev, the first step would be getting it to run (and possibly turn it into a Lasagne Recipe), then we can decide on whether it's good to include in the main library. Ideally, even if only creating a Recipe, it would be very easy to copy/paste.

**Comment by partobs-mdp** [comment]
@f0k Just for the sake of clarity: who is assigned to this issue - me or @botev?

**Comment by f0k** [comment]
> who is assigned to this issue - me or @botev?

Nobody in particular. This is not a quick and easy task, so it doesn't seem useful to block it for a single person. Just use this issue to coordinate -- if you start working on this, leave a short note on what you're doing.

As @botev said (https://github.com/Lasagne/Lasagne/issues/856#issuecomment-312938159), he could use some help in verifying that his implementation is correct. A possible route would be trying to reproduce Figure 5 (left) with the information in Appendices E and F, and create a Lasagne Recipe for that. As a basis, there are Recipes for ResNet, and the DenseNet Recipe uses CIFAR-10.

**Comment by botev** [comment]
I'm not particularly excited to reproduce their experiments, so if @partobs-mdp wants to he can write the recipe for doing that on the CIFAR datasets. The code in the gist is written in the same format as all of Lasagne's optimizers, so just copy pasting it into the recipe should work out of the box. 

Actually, verifying the implementation is correct would be more difficult. Or getting the results from the paper I guess.

**Comment by f0k** [comment]
> as you need to know exactly which variant of the ResNet block the authors are using

True, the paper is not clear about whether they use the variant of the first or second ResNet paper. But at least the code is online at https://github.com/JianGoForIt/YellowFin.

> I have no idea what that means

This part is clear from Table 2 in Appendix E: they use pointwise (1x1) convolutions in the residual blocks. See Figure 5 in the [original ResNet paper](http://arxiv.org/pdf/1512.03385.pdf) if you're interested for more.

But I'd just use CIFAR-10, not CIFAR-100, to keep it simple.

> I personally would opt out for just using the existing Lasagne Recipe for 50 layer ResNet and plug it there.

The ResNet-50 Recipe just shows how to use the ResNet-50 from Caffe's model zoo, it does not have training included. But there's another one for ResNet-56 that includes training on CIFAR-10: https://github.com/Lasagne/Recipes/blob/master/papers/deep_residual_learning/Deep_Residual_Learning_CIFAR-10.py
In addition, there's the DenseNet recipe which also includes training for CIFAR-10: https://github.com/Lasagne/Recipes/tree/master/papers/densenet
And a recipe for the second ResNet paper which includes ResNet-110 and ResNet-164 with preactivation: https://github.com/Lasagne/Recipes/tree/master/papers/preactivation_and_wide_resnet

> Note if the algorithm does not work there at least comparable to Adam, it is either that the implementation has a serious issue (most likely) or the paper results were cherry-picked.

That's a reasonable approach, yes. You could use either of the last three recipes as a starting point and try it with ADAM and your YellowFin implementation. Unfortunately, all three recipes use momentum or Nesterov momentum, not ADAM.

> Actually, verifying the implementation is correct would be more difficult as someone has to sit read the paper routines and steps and verify that they match.

Or compare with their code -- but that may be even more difficult (haven't had a look).

> I've tried in the gist to separate the algorithm into the same routines as described in the paper.

Yes, I've seen and appreciated that! :+1:

**Comment by botev** [comment]
> Unfortunately, all three recipes use momentum or Nesterov momentum, not ADAM.

On that I would highly doubt ADAM to perform significantly worse than Nesterov or normal momentum or if it is that to be significant. But I guess one can run both, as it is just changing one line of code.

**Comment by f0k** [comment]
> On that, I would highly doubt ADAM to perform significantly worse than Nesterov or normal momentum.

... if the learning rate is set well. Since it's invariant to the scale of the gradient, it will need a different learning rate than (Nesterov) momentum. But often the 0.001 default rate works fine.

**Comment by botev** [comment]
Yeah, I've putted the three to run (momentum as from the Recipe, Adam and yellowfin), when ready I will post the trajectories (I'm not sure on K80 how long it takes)

**Comment by botev** [comment]
The results on the training loss are below. Note that all of the algorithms used the same learning rate decay schedule as the original recipe.

![yf](https://user-images.githubusercontent.com/1889878/28152492-b520cbb6-6798-11e7-9a25-ac6f0cad1477.jpg)

The results make things even vaguer, to be honest. The fact that YellowFin works and convergence to the same as momentum suggests that the implementation has a good probability that it is correct, or if it has a mistake it is a very minor, without significant impact, and even harder to detect.

On the other hand, this does not seem to align with the paper results, as there they supposedly even optimised Adam learning rate, while this is just running the default one (again the learning rate is decayed in the same way as the others). Questions, questions... and reproducibility! Does anyone know if Adam in Pytorch/Tensorflow is what we have in Lasagne or maybe they have another one?

**Comment by f0k** [comment]
Interesting!

> The fact that YellowFin works and convergence to the same as momentum suggests that the implementation has a good probability that it is correct

Or at least it produces learning rates / momenta that are not completely off.

> On the other hand, this does not seem to align with the paper results, as there they supposedly even optimised Adam learning rate, while this is just running the default one (again the learning rate is decayed in the same way as the others).

From Appendix F, they used fixed learning rates in their experiments for momentum and ADAM. This raises the question whether YellowFin is only advantageous in their paper because it can decay the learning rate. (For TinyShakespeare and Wall Street Journal, they do decay the learning rate for momentum and ADAM, though.)
Also they initialized the learning rate and momentum for YellowFin to 1.0 and 0.0, while your gist has 0.01 and 0.0 -- what did you use for the experiment?

> Does anyone know if Adam in Pytorch/Tensorflow is what we have in Lasagne or maybe they have another one?

It's identical:
https://github.com/tensorflow/tensorflow/blob/r1.2/tensorflow/python/training/adam.py#L54-L59 (trusting the docstring on this)
https://github.com/Lasagne/Lasagne/blob/master/lasagne/updates.py#L610-L628

> Questions, questions... and reproducibility!

@kastnerkyle: I saw you forked @botev's gist, did you play with it yet? Do you have any results yet?

**Comment by botev** [comment]
> Also they initialized the learning rate and momentum for YellowFin to 1.0 and 0.0, while your gist has 0.01 and 0.0 -- what did you use for the experiment?

In the code I ran I've changed this to `1.0 and 0.0`, while in the gist it is still `0.01`.

> From Appendix F, they used fixed learning rates in their experiments for momentum and ADAM. This raises the question whether YellowFin is only advantageous in their paper because it can decay the learning rate.

I can rerun all of them with fixed learning rates. However, note the learning curves before the decay of the learning rate (about 32k iterations) - YellowFin is still worse. Also by the learning curve, it does seem like the decay helps YellowFin as well. In that regard, I don't really see YellowFin "advantageous" in any way on this specific problem as even if it decays something it does not seem to be doing better than the other algorithms.

**Comment by f0k** [comment]
> However, note the learning curves before the decay of the learning rate (about 32k iterations) - YellowFin is still worse.

True, looking at the paper again, in Figure 5, YellowFin has the lowest training loss right from the start, and keeps it up to 40k iterations on CIFAR-10 with their ResNet variant.

> I can rerun all of them with fixed learning rates.

I don't think that's useful any more, the part before the first learning rate drop is enough.

> Also by the learning curve, it does seem like the decay helps YellowFin as well.

Yes, definitely.

> In that regard, I don't really see YellowFin "advantageous" in any way on this specific problem as even if it decays something it does not seem to be doing better than the other algorithms.

My hypothesis was for the paper, not for your experiment -- that maybe, in Figure 5, YellowFin does better than the others merely because it slightly reduces the learning rate over training time. That wouldn't conflict with the observation that a forced drop helps all algorithms.

> On the other hand, this does not seem to align with the paper results, as there they supposedly even optimised Adam learning rate, while this is just running the default one

From Appendix F, they tried 0.0001, 0.001, 0.01, 0.1 for ADAM and 0.001 performed best, which happens to be the default for ADAM anyway. For SGD, they tried 0.001, 0.01(best), 0.1, 1.0.

Did you monitor which learning rates / momenta YellowFin came up with in some of your experiments?

**Comment by botev** [comment]
> Did you monitor which learning rates / momenta YellowFin came up with in some of your experiments?
No, but I can repeat it and check, should not be too much a hassle.

**Comment by f0k** [comment]
> However, I'm hoping someone to check my code or make more experiments to see how reproducible this is.

The authors already linked your gist in their repository, maybe you can point to our discussion in the comments section of your gist? (Our GPUs are currently all busy, sorry!)

**Comment by botev** [comment]
I'll rerun the same experiment with storing the YellowFin learning rate and momentum as well as with the suggested by the outer smoothing of 0.999. I'll post here the results after it is done.

**Comment by kastnerkyle** [comment]
My GPUs are still busy, but it is on my queue to try when I get one free
(~2-3 more days I hope). One thing that might help is being sure the
initialized weights are *identical*, and compare between PyTorch and
Theano. It will be a pain, but it would eliminate a source of randomness
that can be quite annoying - also be sure the data iteration is identical,
and architecture is identical.

Given these things, it should be possible to look for bugs in the optimizer
more directly, but debugging optimizers is a pain in the neck!

**Comment by botev** [comment]
@kastnerkyle Unfortunately at this moment I don't have the time to do this, as aligning implementations between Theano and Pytorch would be quite hard for me, given that I've never used Pytorch. Additionally, if you expect exactly identical runs you will depend that none of the implementations uses non-deterministic operators. I myself would definitely not be up to making such detailed comparison.

**Comment by kastnerkyle** [comment]
AFAIK Theano and PyTorch are deterministic in their runs (from what ops I
have used at least) given enough seeds are set and so forth. At very least,
I have been able to get repeatable runs in both down to some low eps (1E-6
or more) tolerance. There are ways to make TF behave as well (I believe
reduce_sum is non-deterministic, maybe others), but it is a bit more
involved.

I will report back here when I get a chance to run some code.

**Comment by f0k** [comment]
> AFAIK Theano and PyTorch are deterministic in their runs (from what ops I have used at least)

The backward pass of cuDNN's convolution is nondeterministic unless you explicitly enforce the deterministic implementation (which is slower). I'm not sure how much this affects reproducibility.

> One thing that might help is being sure the initialized weights are *identical*, and compare between PyTorch and Theano.

For now not even the models are identical, it was a quick shot under the assumption that the YellowFin/ADAM comparison would not depend on the exact model being used.
If going that route, it might help to simplify the PyTorch or TF model to a bare minimum and then run it side-by-side with the Theano implementation. It's still a pain.

**Comment by kastnerkyle** [comment]
I was planning to side by side the RNN experiment, to avoid any issues with
pooling, convolution and so on - backwards of max pooling being different
in different frameworks has been an issue for me before. Primarily done
RNNs in PyTorch so far and had good luck with matching architectures +
performance between PyTorch and Theano, however this is probably due to the
simplicity of the ops involved. Looking forward to pain...

**Comment by botev** [comment]
Ok, so now that I fixed that error with running SGD, I guess maybe the implementation is somewhere wrong:

![yf](https://user-images.githubusercontent.com/1889878/28205065-62190024-6878-11e7-8439-a38d9a876c75.jpg)

Here are the learning rate and momentum from YellowFin:

![lr](https://user-images.githubusercontent.com/1889878/28205079-6a26551e-6878-11e7-8e19-6ad0c7d2cd72.jpg)

**Comment by f0k** [comment]
> I guess maybe the implementation is somewhere wrong

Hmm, possibly -- it's striking that it needs so long to catch up with the others. If it wasn't for the learning rate drop, it might have overtaken ADAM, but it's definitely not performing as advertised in Figure 5.

@JianGoForIt: Do the learning rate / momentum trajectories look reasonable to you or very different from your own experiments? (The effect of the learning rate drops enforced from outside is somewhat interesting.)

**Comment by JianGoForIt** [comment]
Hi @f0k,

Thanks for trying out our optimizer. The curve of learning rate and momentum seems strange to me.

Typically for lr and momentum curve, after an initial spike or drop, it should smoothly goes to some relatively smooth level. E.g, snapshot from ResNet 110 on CIFAR10
[fig_hyper_iter_40000.pdf](https://github.com/Lasagne/Lasagne/files/1148009/fig_hyper_iter_40000.pdf)

The initial spike or drop is because we use smoothing on statistics, lr and momentum, which somehow works like an second order differential equation. 

There is another thing to be noticed, if there are some exploding gradient, it might drive the lr and momentum changing crazily(the unstableness in your provided figure might results from this). Some gradient clipping might help. The issues in TF and PyTorch repo has a lot discussions on this. https://github.com/JianGoForIt/YellowFin_Pytorch/issues/1 and 

Another possible issue is the smooth is not strong enough, we uniformly use 0.999 for exponential smoothing in statistics, lr and momentum. Please refer to our PyTorch and TF repo for details.

Please let me know if it helps.

Cheers,

**Comment by botev** [comment]
Just a few extra notes: Notice that the picture in the pdf for the ResNet110 is on log scale, the picture above is in standard linear scale. Second, I still used the `learninig_rate_factor`. I will make one last run well I will try to report everything he has in the pdf and remove the decay factor.

**Comment by botev** [comment]
@JianGoForIt  could you elaborate what the lines in the pdf mean:
lr_min - Is this just the `learning rate`?
dynamic_range - is this `h_min` or `h_max` or?
The others are quite clear.
However, the momentum and learning rate have you plot the solution to the cubic equation or the smoothed versions?

**Comment by JianGoForIt** [comment]
@botev 

lr_min is just the learning rate.
The dynamic range is h_max/h_min.

for the lr and momentum, I think (not 100% sure as the results are from some very old codebase) I plotted the smoothed one fed into the underlying momentum SGD optimizer.

**Comment by botev** [comment]
Ok, I will reproduce the same plots for my experience, but at this stage I'm pretty confident there is an issue somewhere.

**Comment by JianGoForIt** [comment]
@botev Just a quick comment, the smoothing might be trick as there are a lot of them. You might want to refer to our TF repo or PyTorch repo. Cheers.

**Comment by botev** [comment]
@JianGoForIt As a long as you use the same smoothing parameter, as far as I'm aware all of them use exponential moving average and I use smoothing for the following variables:

- h_min
- h_max 
- grad_f (for variance)
- grad_f^2 (for variance)
- ||grad_f||^2 (for distance to optimum)
- ||grad_f|| (for distance to optimum, note that at time `t` this a `sqrt` of the above but I do smoothing average of the `sqrt` rather than the `sqrt`of the above moving average)
- moving average of the ratio of the moving averages for ||grad_f|| and ||grad_f||^2
- alpha
-mu

**Comment by f0k** [comment]
> Typically for lr and momentum curve, after an initial spike or drop, it should smoothly goes to some relatively smooth level.

Note that in @botev's experiment, the learning rate is dropped by a factor of 10 after about 32k and 48k iterations. Except for those disturbances (and the initial spike/drop), the curves are smooth, with slowly increasing momentum and decreasing learning rate as in your plot.

> You might want to refer to our TF repo or PyTorch repo.

Yes, thanks, we'll also compare the code. Thank you for chiming in and sharing your plot!

**Comment by botev** [comment]
Ok, I've re-run one last time YellowFin, without any learning rate decay and reproducing the quantities from the pdf. These are the learning curves:

![yf](https://user-images.githubusercontent.com/1889878/28233968-cfb49ccc-68f2-11e7-87d0-8ca3cf1492c7.jpg)

And this are the quantities of the algorithm (note I'm plotting the smoothed values for all):

![yf_all](https://user-images.githubusercontent.com/1889878/28233978-db545dd8-68f2-11e7-88e2-1871d3fa1487.jpg)

**Comment by JianGoForIt** [comment]
@botev Thanks for the effort. Seems reasonable. Only three comments. The spike for dynamic range is bit large. You might want to check whether turning on and off zero debias has impact on it. 

Regrading decaying you might want also try it for YF, the ideal dropping factor might be different.

My experience is that for resnet like model. Initial momentum 0.9 might be better. It will make the transition phases Short and faster in the beginning. We use default 0.0 because it is uniformly going well for lstm and resent while 0.9 sometimes trigger initial explosion on lstm.

**Comment by JianGoForIt** [comment]
@botev I just noticed there might be an initial loss jump for YF at the very beginning. Could please verify if this is the case? Maybe there is initial explosion of loss at the very beginning.

**Comment by botev** [comment]
@JianGoForIt I think that is happening because the initial learning is a bit too high. I will test it with initial learning rate 0.1 and initial momentum 0.9 which are the one used for the standard momentum algorithm.

**Comment by f0k** [comment]
> I will test it with initial learning rate 0.1 and initial momentum 0.9 which are the one used for the standard momentum algorithm.

Note that there are two possible formulations for the momentum algorithm:
1. velocity := velocity * momentum - eta * grad
2. velocity := velocity * momentum - eta * (1 - momentum) * grad

Lasagne uses the more classical first one, but if I read this code correctly, Torch uses the second one: https://github.com/torch/optim/blob/master/sgd.lua#L62
If that is true, it means hyperparameters in Torch-based papers will not match up with hyperparameters in many other papers (for a momentum of 0.9, the learning rate in a Torch paper will be 10 times as large).
You could try adding `* (1 - updates[mu])` to the learning rate in https://gist.github.com/botev/f8b32c00eafee222e47393f7f0747666#file-yellow_fin-py-L62. Although I'm not sure if YellowFin is actually sensitive to that?

**Comment by botev** [comment]
Well in the paper they use formulation 1. - Equation 1 putting `v_t = x_t - x_{t-1}`. @JianGoForIt could you comment on this

**Comment by JianGoForIt** [comment]
Hi @botev, we use the first formulation here 

velocity := velocity * momentum - eta * grad

**Comment by f0k** [comment]
@JianGoForIt: I see, checking the PyTorch SGD code, they default `dampening` to 0 (https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L52) where the original Torch code defaults `dampening` to `momentum`: https://github.com/torch/optim/blob/master/sgd.lua#L35. Sorry for the red herring.

The docstring in PyTorch [points out another difference](https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L31-L45), though.
Lasagne has:
* velocity := velocity * momentum - eta * grad
* param := param + velocity

PyTorch has:
* velocity := velocity * momentum - grad
* param := param + eta * velocity

Again, not sure if that's critical, given that the Equation 1 in the paper matches Lasagne's update rule (and possibly TensorFlow's as well, haven't checked).

**Comment by cstein06** [comment]
I read the paper carefully and I don't think the results are related with momentum at all. In Figure 4 one sees that the momentum value does not go above 0.8, which shouldn't make much difference (common values are 0.99 for example). 
I confirmed this by turning off momentum and rescaling the learning rate appropriately, getting the same results. Plot:
https://github.com/cstein06/YellowFin/blob/no_momentum/cifar/results/compare_losses.pdf

It means that it is a learning rate scheduler in disguise.

**Comment by botev** [comment]
@cstein06 Could you just let us know which code did you use for the tests. Even so that is an interesting observation.

**Comment by cstein06** [comment]
I ran:
python ./cifar/scripts/CIFAR10-release.py --run_name="no_mom" --mom=0.0
python ./cifar/scripts/CIFAR10-release.py --run_name="with_mom" 
python ./cifar/scripts/compare_no_mom.py

Using this fork: https://github.com/cstein06/YellowFin/tree/no_momentum

The main alteration is in tuner_utils/yellowfin.py, where I use normal SGD, but rescaled:

    if momentum > 0.0:
      print("Using momentum")
      self._optimizer = tf.train.MomentumOptimizer(
        self._lr_var * self.lr_factor, self._mu_var + delta_mu,
        use_locking, name, use_nesterov)
    else:
      # Remove momentum and rescale learning rate accordingly
      print("Not using momentum")
      self._optimizer = tf.train.GradientDescentOptimizer(
        1.0 * self._lr_var * self.lr_factor / (1.0 - self._mu_var))

**Comment by botev** [comment]
Great find! Do you mind running yellowfin with momentum, but not adapting the momentum term (e.g. you will need to change line 295-297 to never change mu). I'm just curious if it somehow keeping momentum is bad for this.

**Comment by cstein06** [comment]
I'm pretty sure the result will be worse, since it is what they compare with in the paper.

**Comment by botev** [comment]
Yes, that's what I meant. Use Yellowfin only for changing the learning rate, not the momentum.

**Comment by mamcn** [comment]
I need different deep reinforcement learning algorithms  convergence rate graph.
thank you!

**Comment by f0k** [comment]
> I need different deep reinforcement learning algorithms convergence rate graph.

Sorry, I'm afraid you are wrong here -- this thread was about implementing a particular research paper (see the very first post) for possible inclusion in Lasagne, a deep learning library, without any connection to reinforcement learning.
Maybe this [reinforcement learning introduction](https://spinningup.openai.com/en/latest/spinningup/bench.html) will help?

**Comment by Sandy4321** [comment]
May you share simple short intro to yellowfin pls

**Comment by f0k** [comment]
> May you share simple short intro to yellowfin pls

Sorry, I'm not aware of any. That's a sign that it hasn't gained a lot of traction. You'll have to read the original paper if you still want to experiment with it.
